pennylane/data/dataset.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the :class:`~pennylane.data.Dataset` class and its associated functions.
+"""
+
+from abc import ABC
+import dill
+import zstd
+
+
+class Dataset(ABC):
+    """Create a dataset object.
+    A dataset might be a collection of information that describes a physical system and its evolution.
+    For example, a dataset for an arbitrary quantum system could have a Hamiltonian, its ground state,
+    and an efficient state-preparation circuit for that state.
+
+    Args:
+        dtype (string): the type of dataset, e.g., `qchem`, `qspin`, etc.
+        **kwargs: variable-length keyworded arguments specifying the data to be stored in the dataset
+
+    .. details::
+        :title: Usage Details
+
+        We can use the :class:`~pennylane.data.Dataset` class to create datasets as follows:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane import numpy as np
+
+            Hamiltonian = qml.Hamiltonian([1., 1.], [qml.PauliZ(wires=0), qml.PauliZ(wires=1)])
+            solution = np.linalg.eigh(qml.matrix(Hamiltonian))
+            ground_energy = np.min(solution[0])
+            ground_state = np.transpose(solution[1])[np.argmin(solution[0])]
```

Comment:
Minor optional suggestion for the variable names: 
Hamiltonian = ...
eigvals, eigvecs = ...
ground_state_energy = ...